### DISCUSSION TITLE
Next Step of Element Plus

### DISCUSSION THREAD

**Discussion by sxzz** [element-plus/element-plus#8288] [Ideas]
✅ = Already done

📝 = Plan

💡 = Idea, further discussion is required

- Workflow
  - 💡 [changesets](https://github.com/changesets/changesets)
    - A tool to manage versioning and changelogs.
    - With it, we can drop [`breakings`](https://github.com/element-plus/element-plus/tree/dev/breakings) folder and avoid to self manage it.
  - 💡 GitHub flow. Only one main branch.
    - Rename `dev` branch to `main`.
    - Remove `master` branch.
    - Reason: I think GitHub flow is more suitable for us, a open-source project. Like vue, ant-design...
    - RFC: https://github.com/element-plus/element-plus/discussions/8439
  - Version Publishing
    - ✅ Use GitHub Actions to publish npm packages.
    - 💡 Only create a new tag, to automatically publish a new version without writing changelog manually.
      - 💡 The changelog will generated by` changesets` with two languages (English and Chinese)
- Project
  - 📝 A real monorepo
    - Currently Element Plus uses the pnpm workspace to manage the entire project, but does not publish all the packages separately in the npm registry.
    - We should publish all packages to npm registry, and the developers can freely choose the parts they want to use.
  - 💡 Greater extensibility
    - Expose more general functions, and move the logic of view model to a separate package.
    - Goal: make more versions of Element Plus possible. Like React, Vue 2 or Svelte.
  - 📝 Use dayjs 2
    - Supports both CommonJS and ESM.
  - 💡 Extract icons
    - User can customize bulit-in icons.
    - Provide icons when install
    - #8437
- Build
  - 📝 Optimize compilation speed
    - Use esbuild to bundle. A prototype repo is available here for demonstration purposes.
    - https://github.com/sxzz/element-plus-next
  - 📝 Drop gulp.js
    - No maintenance for a long time
    - use `pnpm run --parallel`

For preview: https://github.com/sxzz/element-plus-next

🌟 More ideas and discussions are welcome.

**Comment by xxholly32** [comment]
Are there any plans for [element-plus-next](https://github.com/sxzz/element-plus-next) in the future?

I created my own atomic project, <https://xxholly32.github.io/element-plus-pro/>, and I wanted the components, documentation, and tools to be more atomized. I really wish I could start by separating packages.

**Reply by ryuhangyeong** [reply]
@xxholly32 

https://github.com/element-plus/element-plus/pull/11720

https://github.com/element-plus/element-plus/tree/next

**Reply by wzc520pyfm** [reply]
Do we have a development plan for the next version? I can help!

**Comment by xxholly32** [comment]
Do you have any plans to release 3.0.0-alpha ？I expect it so much

**Comment by gimjin** [comment]
recommended to use prop to implement the event, because vue3 removes emit from the core code base, and only keeps it for compatibility with vue2. Prop has more advantages in implementing events, such as passing parameters, asynchronous function execution, etc.

like el-upload on-*
